draft-ietf-quic-version-negotiation.md

```diff
@@ -392,13 +392,32 @@ its length is not divisible by four), then the endpoint MUST close the
 connection; if the connection was using QUIC version 1, that connection closure
 MUST use a transport error of type TRANSPORT_PARAMETER_ERROR. If an endpoint
 receives a Chosen Version equal to zero, or any Available Version equal to zero,
-it MUST treat it as a parsing failure.
+it MUST treat it as a parsing failure. If a server receives a Version
+Information where the Chosen Version is not included in Available Versions, it
+MUST treat it as a parsing failure.
 
 Every QUIC version that supports version negotiation MUST define a method for
 closing the connection with a version negotiation error. For QUIC version 1,
 version negotiation errors are signaled using a transport error of type
 VERSION_NEGOTIATION_ERROR; see {{iana-error}}.
 
+When a server receives a client's first flight, the server will first establish
+which QUIC version is in use for this connection in order to properly parse the
+first flight. For example, QUIC version 1 is in use when the Version field of
+the Long Headers that are part of the client's first flight is set to
+0x00000001. When the server then parses the client's Version Information, the
+server MUST validate that the client's Chosen Version matches the version in use
+for the connection. If the two differ, the server MUST close the connection with
```

Comment:
Not sure if this is going to be correct in all cases.

The client might switch to using Version 2 if the server decides to upgrade.  Retransmissions of the ClientHello containing the Version Information might then be sent in Initial packets with a version == 2.

Now, obviously, the server can't upgrade until it has read the Version Information, but there is a possibility that a client could retransmit stuff that the server already has.  This might be due to the Version Information being in a packet that the server received, processed, and acknowledged.  But the ACK might get lost and so the client might still end up sending another copy.

Also, I don't believe that this sort of validation is easy to implement.  The principle we apply here is that validation is mandated only where there is fairly uniform agreement that it is trivial to implement.  This is not.  It requires accessing information that is very hard to access in our stack.  In our stack, processing of Version Information occurs in a callback that is invoked by the TLS library.  For various reasons (mostly relating to how callbacks from C work with Rust and Rust's ownership rules), that processing cannot access a lot of connection-level state.  It certainly can't access details of the packet that was received, even if that would make sense.  After all, the data that the cryptographic handshake processes could come from multiple packets, some of which might have been received and long-since discarded.

Reply:
The text above along discusses the client's first flight. The retransmission scenario would not be part of the first flight, by definition. The validation is trivial to implement because you have access to the version of the connection - we're not asking to keep track of the version of the packet carrying the CRYPTO frame, only the version of the first packet received. Your stack already has access to this during TP parsing because the TP parsing is version-dependent.

Reply:
But there is no guarantee that the version information is in that first packet, so this is not that easy to implement.